keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2346 @@
+
+If a Pod references multiple claims managed by the same driver, then the driver
+can combine updating `UnsuitableNodes` for all of them, after considering all
+claims.
+
+### API
+
+The PodSpec gets extended. Types and structs referenced from PodSpec
+(ResourceClaimTemplate, ResourceClaimSpec) must be placed in `core.k8s.io/v1`
+to keep that package self-contained. The new fields in the PodSpec are gated by
+the DynamicResourceAllocation feature gate and can only be set when it is
+enabled. Initially, they are declared as alpha and then can still be changed
+from one release to the next.
+
+ResourceClaim and ResourceClass are new built-in types in `core.k8s.io/v1`.
+They only get registered when the feature gate is enabled.
+
+Putting everything into `core.k8s.io/v1` was chosen instead of using CRDs
+because core Kubernetes components must interact with the new objects and
+installation of CRDs as part of cluster creation is an unsolved problem. They
+could be defined in some other API group (`node.k8s.io`, some new group) but
+having everything in the same group is less confusing for users.
+
+Secrets are not part of this API: if a resource driver needs secrets, for
+example to access its own backplane, then it can define custom parameters for
+those secrets and retrieve them directly from the apiserver. This works because
+drivers are expected to be written for Kubernetes.
+
+```
+// ResourceClass is used by administrators to influence how resources
+// are allocated.
+type ResourceClass struct {
+	metav1.TypeMeta
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta
+
+	// DriverName determines which resource driver is to be used for
+	// allocation of a ResourceClaim that uses this class.
+	//
+	// Resource drivers have a unique name in normal domain order
+	// (acme.example.com).
+	DriverName string
+
+	// Parameters references an arbitrary separate object that may hold
+	// parameters that will be used by the
+	// driver when allocating a resource that uses this class. The driver
+	// will be able to distinguish between parameters stored here and and
+	// those stored in ResourceClaimSpec. These parameters here can only be
+	// set by cluster administrators.
+	//
+	// The object must be cluster-scoped.
+	Parameters ResourceClassParametersReference
+
+	// Only nodes matching the selector will be considered by the scheduler
+	// when trying to find a Node that fits a Pod when that Pod uses
+	// a ResourceClaim that has not been allocated yet.
+	//
+	// Setting this field is optional. If nil, all nodes are candidates.
+	SuitableNodes *core.NodeSelector
+}
+
+// ResourceClaim is created by users to describe which resources they need.
+// Its status tracks whether the resource has been allocated and what the
+// resulting attributes are.
+type ResourceClaim struct {
+	metav1.TypeMeta
+
+	// The driver must set a finalizer here before it attempts to allocate
```

Comment:
It answers one question that a user might have: "may I `kubectl delete` my ResourceClaim at any time?

I don't feel strongly about this. I agree about the incomplete protection (but see the other comment about "complete protection" not being possible in a KEP because it'll depend on cloud vendor specific features).

As user-facing documentation it may be worth adding "never remove finalizers just to get a ResourceClaim removed, this may cause a resource leak".

Reply:
We can deal with comments at actual API review time :)